{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def denovo_links(bam_file, out_name):\n",
    "    \"\"\"\n",
    "    Function to be applied to sorted bam file\n",
    "    \"\"\"\n",
    "    print('Starting ', bam_file)\n",
    "    output_file = open(out_name,'w')\n",
    "    \n",
    "    samfile = pysam.AlignmentFile(bam_file, 'rb')\n",
    "    count = 0\n",
    "    genomes = {'pB10::rfp_putative':'pB10', 'NC_000913.3_gfp':'E.Coli'}\n",
    "    pB10_pB10 = 0\n",
    "    pB10_ecoli = 0\n",
    "    ecoli_ecoli = 0\n",
    "    pB10_unaligned = 0\n",
    "    unaligned_unaligned = 0\n",
    "    multimapped = 0\n",
    "    \n",
    "    pairs = []\n",
    "    \n",
    "    #lists that store the location on the pB10 genome of each aligned read\n",
    "    pB10_pB10_loc = []\n",
    "    pB10_unknown_loc = []\n",
    "    pB10_ecoli_loc = []\n",
    "    \n",
    "    global p_unaligned\n",
    "    global p_p\n",
    "    global p_ecoli\n",
    "    p_unaligned = [0]*pb10_len\n",
    "    p_p = [0]*pb10_len\n",
    "    p_ecoli = [0]*pb10_len\n",
    "    \n",
    "    info_df = pd.DataFrame(columns = ['seqid', 'pB10_start','read_length'])\n",
    "\n",
    "    for read in samfile.fetch(until_eof=True):\n",
    "        count += 1\n",
    "        genome = ''\n",
    "        name = read.query_name\n",
    "        cigar = ''\n",
    "        seq = ''\n",
    "        loc = ''\n",
    "        if read.is_unmapped:\n",
    "            genome = 'unaligned'\n",
    "            cigar = read.cigarstring\n",
    "            seq = read.query_sequence\n",
    "            length = len(seq)\n",
    "        if not read.is_unmapped:\n",
    "            genome = genomes.get(read.reference_name)\n",
    "            cigar = read.cigarstring\n",
    "            seq = read.query_sequence\n",
    "            match = str(len(seq))+'M'\n",
    "            loc = read.reference_start\n",
    "            length = len(seq)\n",
    "            if cigar != match:\n",
    "                genome = 'unaligned'\n",
    "        pairs.append([name, genome, loc, length, seq, read.get_tags()])\n",
    "        if count % 2 == 0:\n",
    "            r1_multimapped = check_alt(pairs[0][1], pairs[0][5], pairs[0][3])\n",
    "            r2_multimapped = check_alt(pairs[1][1], pairs[1][5], pairs[1][3])\n",
    "            if len(r1_multimapped) > 1 or len(r2_multimapped) >1:\n",
    "                multimapped += 1\n",
    "                pairs = []\n",
    "            else:\n",
    "                links = pairs[0][1] + pairs[1][1]\n",
    "                if links == 'pB10pB10':\n",
    "                    pB10_pB10 += 1\n",
    "                    update_counts('plasmid','plasmid', [pairs[0][2], pairs[1][2]], [pairs[0][3], pairs[1][3]])\n",
    "                    pB10_pB10_loc.append(pairs[0][2])\n",
    "                    pB10_pB10_loc.append(pairs[1][2])\n",
    "                if links == 'pB10unaligned':\n",
    "                    pB10_unaligned += 1\n",
    "                    update_counts('plasmid', 'unaligned', pairs[0][2], pairs[0][3])\n",
    "                    pB10_unknown_loc.append(pairs[0][2])\n",
    "                    info_df = info_df.append({'seqid':pairs[1][0], 'pB10_start':pairs[0][2], 'read_length':pairs[0][3]}, ignore_index=True)\n",
    "                    output_file.write(\">\" + pairs[1][0] + '\\n' + pairs[1][4] + '\\n')\n",
    "                if links == 'unalignedpB10':\n",
    "                    pB10_unaligned += 1\n",
    "                    update_counts('unaligned', 'plasmid', pairs[1][2], pairs[1][3])\n",
    "                    pB10_unknown_loc.append(pairs[1][2])\n",
    "                    info_df = info_df.append({'seqid':pairs[0][0], 'pB10_start':pairs[1][2], 'read_length':pairs[1][3]}, ignore_index=True)\n",
    "                    output_file.write(\">\" + pairs[0][0] + '\\n' + pairs[0][4] + '\\n')\n",
    "                if links == 'pB10E.Coli':\n",
    "                    pB10_ecoli += 1\n",
    "                    update_counts('plasmid', 'ecoli', pairs[0][2], pairs[0][3])\n",
    "                    pB10_ecoli_loc.append(pairs[0][2])\n",
    "                if links == 'E.ColipB10':\n",
    "                    pB10_ecoli += 1\n",
    "                    update_counts('ecoli', 'plasmid', pairs[1][2], pairs[1][3])\n",
    "                    pB10_ecoli_loc.append(pairs[1][2])\n",
    "                if links == 'E.ColiE.Coli':\n",
    "                    ecoli_ecoli += 1\n",
    "                if links == 'unalignedunaligned':\n",
    "                    unaligned_unaligned += 1\n",
    "                pairs = []\n",
    "    output_file.close()\n",
    "    print(bam_file, 'Completed')\n",
    "    print(np.array([count/2, pB10_pB10, pB10_ecoli, pB10_unaligned, unaligned_unaligned, multimapped], '\\n'))\n",
    "    return(np.array([[count/2, pB10_pB10, pB10_ecoli, pB10_unaligned, unaligned_unaligned, multimapped],\n",
    "                     [p_unaligned, p_p, p_ecoli],\n",
    "                     [pB10_unknown_loc, pB10_pB10_loc, pB10_ecoli_loc]]), \n",
    "           info_df)\n",
    "\n",
    "def update_counts(ref, link, start_index, length):\n",
    "    \"\"\"Function for updating the coverage list\"\"\"\n",
    "    global p_putida\n",
    "    global p_ecoli\n",
    "    global p_p\n",
    "    \n",
    "    if ref == 'plasmid':\n",
    "        if link == 'ecoli':\n",
    "            s = start_index\n",
    "            e = start_index + length\n",
    "            for i in range(s, e):\n",
    "                if i < pb10_len:\n",
    "                    p_ecoli[i] += 1\n",
    "                else:\n",
    "                    i = i - pb10_len\n",
    "                    p_ecoli[i] += 1 \n",
    "\n",
    "    if link == 'plasmid':\n",
    "        if ref == 'ecoli':\n",
    "            s = start_index\n",
    "            e = start_index + length\n",
    "            for i in range(s, e):\n",
    "                if i < pb10_len:\n",
    "                    p_ecoli[i] += 1\n",
    "                else:\n",
    "                    i = i - pb10_len\n",
    "                    p_ecoli[i] += 1 \n",
    "                    \n",
    "    if ref == 'plasmid' and link == 'plasmid':                  \n",
    "        for i in range(0,2):\n",
    "            s = start_index[i]\n",
    "            e = start_index[i] + length[i]\n",
    "            for i in range(s, e):\n",
    "                if i < pb10_len:\n",
    "                    p_p[i] += 1\n",
    "                else:\n",
    "                    i = i - pb10_len\n",
    "                    p_p[i] += 1\n",
    "                    \n",
    "def check_alt(aligned, tags, length):\n",
    "    \"\"\"\n",
    "    This function uses the alignment tags to determine if a read was\n",
    "    mapped perfectly to multiple references. The output is a list of \n",
    "    references to which the read mapped.\n",
    "    \"\"\"\n",
    "    genomes = [aligned]\n",
    "    for tag in tags:\n",
    "        if tag[0] == 'XA' or tag[0] == 'SA':\n",
    "            tag = list(tag)\n",
    "            others = tag[1].split(';')[:-1]\n",
    "            fmatch = []\n",
    "            for alt in others:\n",
    "                info = alt.split(',')\n",
    "                if info[-2] == str(length)+'M':\n",
    "                    fmatch.append(info[0])\n",
    "            for i in fmatch:\n",
    "                genomes.append(i)\n",
    "    return(list(set(genomes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting  C+_aligned_sorted.bam\n"
     ]
    }
   ],
   "source": [
    "os.chdir('/mnt/ceph/cast9836/00_projects/hic_targetcapture/05_bwa_alignment/de_novo/')\n",
    "\n",
    "pb10_len = 68345\n",
    "p_p = [0]*pb10_len\n",
    "p_ecoli = [0]*pb10_len\n",
    "p_putida = [0]*pb10_len\n",
    "\n",
    "Bplus_allden, Bplus_df = denovo_links('C+_aligned_sorted.bam', '../../06_output/C+_pothosts.fasta')\n",
    "#Cplus_allden, Cplus_df = denovo_links('D+_aligned_sorted.bam', '../../06_output/D+_pothosts.fasta')\n",
    "#Dplus_allden, Dplus_df = denovo_links('E+_aligned_sorted.bam', '../../06_output/E+_pothosts.fasta')\n",
    "#Eplus_allden, Eplus_df = denovo_links('F+_aligned_sorted.bam', '../../06_output/F+_pothosts.fasta')\n",
    "#Fplus_allden, Fplus_df = denovo_links('G+_aligned_sorted.bam', '../../06_output/G+_pothosts.fasta')\n",
    "#ctrl1plus_allden, ctrlplus_df = denovo_links('ctrl1+_ecoli_aligned_sorted.bam', '../../06_output/de_novo/ctrl1+_pothosts.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Bplus_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-33-ac3ca1e68655>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mBplus_df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'Bplus_df' is not defined"
     ]
    }
   ],
   "source": [
    "Bplus_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "os.chdir('/mnt/ceph/cast9836/00_projects/hic_targetcapture/06_output/de_novo/')\n",
    "\n",
    "cplus_taxon = pd.read_csv('C+_lineage.txt', sep='\\t')\n",
    "cplus_kraken = pd.read_csv('C+_out.kraken', sep='\\t', header=None)\n",
    "cplus_taxon['seqid'] = cplus_kraken[1]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
