{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.gridspec as gridspec\n",
    "from matplotlib.offsetbox import AnchoredText\n",
    "import matplotlib.ticker as ticker\n",
    "import copy\n",
    "import random\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_links(bam_file):\n",
    "    \"\"\"\n",
    "    Function to be applied to file after matlock filtering, no unaligned reads should be present.\n",
    "    Takes a bam file as input and returns the counts for the links present.\n",
    "    \"\"\"\n",
    "    print('Starting ', bam_file)\n",
    "    samfile = pysam.AlignmentFile(bam_file, 'rb')\n",
    "    count = 0\n",
    "    genomes = {'pB10::rfp_putative':'pB10', 'NC_002947.4':'P.Putida', 'NC_000913.3_gfp':'E.Coli'}\n",
    "    pB10_pB10 = 0\n",
    "    pB10_ecoli = 0\n",
    "    pB10_pputida = 0\n",
    "    ecoli_ecoli = 0\n",
    "    pputida_pputida = 0\n",
    "    multimapped = 0\n",
    "    pairs = []\n",
    "    \n",
    "    #ecoli_file = open('/mnt/ceph/cast9836/00_projects/hic_targetcapture/06_output/'+bam_file+'ecoli_pb10_chromosome.fasta','w')\n",
    "    #pputida_file = open('/mnt/ceph/cast9836/00_projects/hic_targetcapture/06_output/'+bam_file+'pputida_pb10_chromosome.fasta','w')\n",
    "    \n",
    "    #lists that store the location on the pB10 genome of each aligned read\n",
    "    pB10_pB10_loc = []\n",
    "    pB10_pputida_loc = []\n",
    "    pB10_ecoli_loc = []\n",
    "    \n",
    "    #lists that store location on chromosome of pb10 aligned reads\n",
    "    ecoli_pB10_loc = []\n",
    "    pputida_pB10_loc = []\n",
    "    \n",
    "    global p_putida\n",
    "    global p_p\n",
    "    global p_ecoli\n",
    "    p_putida = [0]*pb10_len\n",
    "    p_p = [0]*pb10_len\n",
    "    p_ecoli = [0]*pb10_len\n",
    "    \n",
    "    for read in samfile.fetch(until_eof=True):\n",
    "        count += 1\n",
    "        genome = ''\n",
    "        name = ''\n",
    "        cigar = ''\n",
    "        loc = ''\n",
    "        seq = read.query_sequence\n",
    "        length = len(seq)\n",
    "        if read.is_unmapped:\n",
    "            name = read.query_name\n",
    "            genome = 'unaligned'\n",
    "        if not read.is_unmapped:\n",
    "            name = read.query_name\n",
    "            genome = genomes.get(read.reference_name)\n",
    "            cigar = read.cigarstring\n",
    "            loc = read.reference_start\n",
    "            #match = str(length)+'M'\n",
    "            #if cigar != match:\n",
    "            #    genome='unaligned'\n",
    "        pairs.append([name, genome, loc, length, read.get_tags(), read.query_sequence])\n",
    "        if count % 2 == 0:\n",
    "            r1_multimapped = check_alt(pairs[0][1], pairs[0][4], pairs[0][3])\n",
    "            r2_multimapped = check_alt(pairs[1][1], pairs[1][4], pairs[1][3])\n",
    "            if len(r1_multimapped) > 1 or len(r2_multimapped) >1:\n",
    "                multimapped += 1\n",
    "                pairs = []\n",
    "            else:\n",
    "                links = pairs[0][1] + pairs[1][1]\n",
    "                if links == 'pB10pB10':\n",
    "                    pB10_pB10 += 1\n",
    "                    update_counts('plasmid','plasmid', [pairs[0][2], pairs[1][2]], [pairs[0][3], pairs[1][3]])\n",
    "                    pB10_pB10_loc.append(pairs[0][2])\n",
    "                    pB10_pB10_loc.append(pairs[1][2])\n",
    "                if links == 'pB10P.Putida':\n",
    "                    pB10_pputida += 1\n",
    "                    update_counts('plasmid', 'putida', pairs[0][2], pairs[0][3])\n",
    "                    pB10_pputida_loc.append(pairs[0][2])\n",
    "                    pputida_pB10_loc.append(pairs[1][2])\n",
    "                    #pputida_file.write(\">\" + pairs[1][0] + '\\n' + pairs[1][-1] + '\\n')\n",
    "                if links == 'P.PutidapB10':\n",
    "                    pB10_pputida += 1\n",
    "                    update_counts('putida', 'plasmid', pairs[1][2], pairs[1][3])\n",
    "                    pB10_pputida_loc.append(pairs[1][2])\n",
    "                    pputida_pB10_loc.append(pairs[0][2])\n",
    "                    #pputida_file.write(\">\" + pairs[0][0] + '\\n' + pairs[0][-1] + '\\n')\n",
    "                if links == 'pB10E.Coli':\n",
    "                    pB10_ecoli += 1\n",
    "                    update_counts('plasmid', 'ecoli', pairs[0][2], pairs[0][3])\n",
    "                    pB10_ecoli_loc.append(pairs[0][2])\n",
    "                    ecoli_pB10_loc.append(pairs[1][2])\n",
    "                    #ecoli_file.write(\">\" + pairs[1][0] + '\\n' + pairs[1][-1] + '\\n')\n",
    "                if links == 'E.ColipB10':\n",
    "                    pB10_ecoli += 1\n",
    "                    update_counts('ecoli', 'plasmid', pairs[1][2], pairs[1][3])\n",
    "                    pB10_ecoli_loc.append(pairs[1][2])\n",
    "                    ecoli_pB10_loc.append(pairs[1][2])\n",
    "                    #ecoli_file.write(\">\" + pairs[0][0] + '\\n' + pairs[0][-1] + '\\n')\n",
    "                if links == 'E.ColiE.Coli':\n",
    "                    ecoli_ecoli += 1 \n",
    "                if links == 'P.PutidaP.Putida':\n",
    "                    pputida_pputida += 1\n",
    "                pairs = []\n",
    "    return(np.array([[count/2, pB10_pB10, pB10_ecoli, pB10_pputida],\n",
    "                    [p_putida, p_p, p_ecoli], \n",
    "                    [pB10_pputida_loc, pB10_pB10_loc, pB10_ecoli_loc, ecoli_pB10_loc, pputida_pB10_loc]]))\n",
    "\n",
    "def update_counts(ref, link, start_index, length):\n",
    "    \"\"\"Function for updating the coverage list\"\"\"\n",
    "    global p_putida\n",
    "    global p_ecoli\n",
    "    global p_p\n",
    "    \n",
    "    if ref == 'plasmid':\n",
    "        if link == 'ecoli':\n",
    "            s = start_index\n",
    "            e = start_index + length\n",
    "            for i in range(s, e):\n",
    "                if i < pb10_len:\n",
    "                    p_ecoli[i] += 1\n",
    "                else:\n",
    "                    i = i - pb10_len\n",
    "                    p_ecoli[i] += 1 \n",
    "        if link == 'putida':\n",
    "            s = start_index\n",
    "            e = start_index + length\n",
    "            for i in range(s, e):\n",
    "                if i < pb10_len:\n",
    "                    p_putida[i] += 1\n",
    "                else:\n",
    "                    i = i - pb10_len\n",
    "                    p_putida[i] += 1 \n",
    "        \n",
    "\n",
    "    if link == 'plasmid':\n",
    "        if ref == 'ecoli':\n",
    "            s = start_index\n",
    "            e = start_index + length\n",
    "            for i in range(s, e):\n",
    "                if i < pb10_len:\n",
    "                    p_ecoli[i] += 1\n",
    "                else:\n",
    "                    i = i - pb10_len\n",
    "                    p_ecoli[i] += 1 \n",
    "        if ref == 'putida':\n",
    "            s = start_index\n",
    "            e = start_index + length\n",
    "            for i in range(s, e):\n",
    "                if i < pb10_len:\n",
    "                    p_putida[i] += 1\n",
    "                else:\n",
    "                    i = i - pb10_len\n",
    "                    p_putida[i] += 1 \n",
    "                    \n",
    "    if ref == 'plasmid' and link == 'plasmid':                  \n",
    "        for i in range(0,2):\n",
    "            s = start_index[i]\n",
    "            e = start_index[i] + length[i]\n",
    "            for i in range(s, e):\n",
    "                if i < pb10_len:\n",
    "                    p_p[i] += 1\n",
    "                else:\n",
    "                    i = i - pb10_len\n",
    "                    p_p[i] += 1\n",
    "                    \n",
    "def check_alt(aligned, tags, length):\n",
    "    \"\"\"\n",
    "    This function uses the alignment tags to determine if a read was\n",
    "    mapped perfectly to multiple references. The output is a list of \n",
    "    references to which the read mapped.\n",
    "    \"\"\"\n",
    "    genomes = [aligned]\n",
    "    for tag in tags:\n",
    "        if tag[0] == 'XA' or tag[0] == 'SA':\n",
    "            tag = list(tag)\n",
    "            others = tag[1].split(';')[:-1]\n",
    "            fmatch = []\n",
    "            for alt in others:\n",
    "                info = alt.split(',')\n",
    "                if info[-2] == str(length)+'M':\n",
    "                    fmatch.append(info[0])\n",
    "            for i in fmatch:\n",
    "                genomes.append(i)\n",
    "    return(list(set(genomes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting  B_sorted_aligned.bam\n"
     ]
    }
   ],
   "source": [
    "os.chdir('/mnt/ceph/cast9836/00_projects/hic_targetcapture/05_bwa_alignments/enrichment_comparison/sorted/')\n",
    "\n",
    "pb10_len = 68345\n",
    "p_p = [0]*pb10_len\n",
    "p_ecoli = [0]*pb10_len\n",
    "p_putida = [0]*pb10_len\n",
    "\n",
    "A_all = extract_links('B_sorted_aligned.bam')\n",
    "#B_all = extract_links('C_sorted_aligned.bam')\n",
    "#C_all = extract_links('D_sorted_aligned.bam')\n",
    "#D_all = extract_links('E_sorted_aligned.bam')\n",
    "#E_all = extract_links('F_sorted_aligned.bam')\n",
    "#ctrl1_all = extract_links('ctrl1_ecoli_sorted_aligned.bam')\n",
    "\n",
    "#Bplus_all = extract_links('C+_sorted_aligned.bam')\n",
    "#Cplus_all = extract_links('D+_sorted_aligned.bam')\n",
    "#Dplus_all = extract_links('E+_sorted_aligned.bam')\n",
    "#Eplus_all = extract_links('F+_sorted_aligned.bam')\n",
    "#Fplus_all = extract_links('G+_sorted_aligned.bam')\n",
    "#ctrl1plus_all = extract_links('ctrl1+_ecoli_sorted_aligned.bam')\n",
    "#ctrl2plus_all = extract_links('ctrl2+_soil_sorted_aligned.bam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_count = pd.DataFrame({'pb10_pb10': [A_all[0][1]],\n",
    "                           'pb10_ecoli': [A_all[0][2]],\n",
    "                           'pb10_pputida': [A_all[0][3]]})\n",
    "plasmid_coverage = pd.DataFrame({'pb10_pb10':A_all[1][0], \n",
    "                                 'pb10_ecoli':A_all[1][1], \n",
    "                                 'pb10_pputida':A_all[1][2]})\n",
    "with open('test.txt', 'w') as file:\n",
    "    file.write('pB10_pputida_loc, pB10_pB10_loc, pB10_ecoli_loc, ecoli_pB10_loc, pputida_pB10_loc\\n')\n",
    "    for item in A_all[2]:\n",
    "        file.write(str(item))\n",
    "        file.write('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'B'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = 'B_xyz'\n",
    "test.split('_')[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get sequencing depth for each library from multiqc report\n",
    "os.chdir('/mnt/ceph/cast9836/00_projects/hic_targetcapture/02_fastqc/untrimmed/multiqc_data/')\n",
    "seq_depth = pd.read_csv('multiqc_general_stats.txt', sep='\\t')\n",
    "seq_depth = seq_depth[['Sample', 'FastQC_mqc-generalstats-fastqc-total_sequences']]\n",
    "seq_depth.rename(columns={'FastQC_mqc-generalstats-fastqc-total_sequences':'num_reads'}, inplace=True)\n",
    "seq_depth = seq_depth.iloc[::2]\n",
    "seq_depth['Sample'] = seq_depth.Sample.str[:-3]\n",
    "seq_depth.reset_index(inplace=True)\n",
    "seq_depth.drop(columns=['index'], inplace=True)\n",
    "seq_depth['Type'] = ['Hi-C', 'Hi-C+', 'Hi-C', 'Hi-C+', 'Hi-C', 'Hi-C+', 'Hi-C', 'Hi-C+', 'Hi-C', 'Hi-C+', 'Hi-C+', 'Hi-C', 'Hi-C+']\n",
    "seq_depth['Sample'] = ['A', 'B', 'B', 'C', 'C', 'D', 'D', 'E', 'E', 'F', 'Ctrl1', 'Ctrl1', 'Ctrl2']\n",
    "seq_depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#place hic-data into dataframe, output two csv files\n",
    "#first is raw counts and second is normalized per million counts\n",
    "hic_all = pd.DataFrame([A_all[0], B_all[0], C_all[0], D_all[0], E_all[0], ctrl1_all[0]],\n",
    "                     columns = ['count', 'multimapped', 'pB10_pB10','pB10_ecoli','pB10_pputida'])\n",
    "hic_all['Sample'] = ['A', 'B', 'C', 'D', 'E', 'Ctrl1']\n",
    "hic_all = hic_all.merge(seq_depth[seq_depth['Type']=='Hi-C']).drop(['count', 'Type'], axis=1)\n",
    "hic_all.set_index(\"Sample\", inplace=True)\n",
    "hic_all.to_csv('/mnt/ceph/cast9836/00_projects/hic_targetcapture/06_output/hic_counts_raw.csv')\n",
    "\n",
    "hic_all['standard'] = seq_depth['num_reads'].min()/hic_all['num_reads']\n",
    "hic_all.drop(columns='num_reads', inplace=True)\n",
    "\n",
    "#normalize hic counts\n",
    "hic_all['pB10_pB10'] = hic_all['pB10_pB10']*hic_all['standard']\n",
    "hic_all['pB10_ecoli'] = hic_all['pB10_ecoli']*hic_all['standard']\n",
    "hic_all['pB10_pputida'] = hic_all['pB10_pputida']*hic_all['standard']\n",
    "\n",
    "hic_all.drop(columns='standard', inplace=True)\n",
    "hic_all.to_csv('/mnt/ceph/cast9836/00_projects/hic_targetcapture/06_output/hic_counts_normalized.csv')\n",
    "\n",
    "#place hic+-data into dataframe, output two csv files\n",
    "#first is raw counts and second is normalized per million counts\n",
    "hicplus_all = pd.DataFrame([Bplus_all[0], Cplus_all[0], Dplus_all[0], Eplus_all[0], Fplus_all[0], ctrl1plus_all[0], ctrl2plus_all[0]],\n",
    "                     columns = ['count', 'multimapped', 'pB10_pB10','pB10_ecoli','pB10_pputida'])\n",
    "hicplus_all['Sample'] = ['B', 'C', 'D', 'E', 'F', 'Ctrl1', 'Ctrl2']\n",
    "hicplus_all = hicplus_all.merge(seq_depth[seq_depth['Type']=='Hi-C+']).drop(['count', 'Type'], axis=1)\n",
    "hicplus_all.set_index(\"Sample\", inplace=True)\n",
    "hicplus_all.to_csv('/mnt/ceph/cast9836/00_projects/hic_targetcapture/06_output/hic+_counts_raw.csv')\n",
    "\n",
    "hicplus_all['standard'] = seq_depth['num_reads'].min()/hicplus_all['num_reads']\n",
    "norm_factor = list(hicplus_all['standard'])\n",
    "print(hicplus_all)\n",
    "hicplus_all.drop(columns='num_reads', inplace=True)\n",
    "\n",
    "#normalize hic counts\n",
    "hicplus_all['pB10_pB10'] = hicplus_all['pB10_pB10']*hicplus_all['standard']\n",
    "hicplus_all['pB10_ecoli'] = hicplus_all['pB10_ecoli']*hicplus_all['standard']\n",
    "hicplus_all['pB10_pputida'] = hicplus_all['pB10_pputida']*hicplus_all['standard']\n",
    "\n",
    "hicplus_all.drop(columns='standard', inplace=True)\n",
    "hicplus_all.to_csv('/mnt/ceph/cast9836/00_projects/hic_targetcapture/06_output/hic+_counts_normalized.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(hic_all)\n",
    "print(hicplus_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generation of Figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare dataframe for plotting\n",
    "pr_hic_all = {'Index':hic_all.index.tolist()*3, 'Hi-C link':['pB10-pB10']*6 + ['Ecoli-pB10']*6 + ['Pputida-pB10']*6, \n",
    "      'count':hic_all['pB10_pB10'].tolist()+hic_all['pB10_ecoli'].tolist()+hic_all['pB10_pputida'].tolist()}\n",
    "plot_ready_hic_all = pd.DataFrame.from_dict(pr_hic_all)\n",
    "plot_ready_hic_all['type'] = 'Hi-C'\n",
    "\n",
    "pr_hicplus_all = {'Index':hicplus_all.index.tolist()*3, 'Hi-C link':['pB10-pB10']*7 + ['Ecoli-pB10']*7 + ['Pputida-pB10']*7, \n",
    "      'count':hicplus_all['pB10_pB10'].tolist()+hicplus_all['pB10_ecoli'].tolist()+hicplus_all['pB10_pputida'].tolist()}\n",
    "plot_ready_hicplus_all = pd.DataFrame.from_dict(pr_hicplus_all)\n",
    "plot_ready_hicplus_all['type'] = 'Hi-C+'\n",
    "all_data = pd.concat([plot_ready_hic_all, plot_ready_hicplus_all], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure1data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/mnt/ceph/cast9836/00_projects/hic_targetcapture/06_output/figures/')\n",
    "figure1data = all_data.loc[(all_data['type']=='Hi-C') & \n",
    "                           (all_data['Hi-C link']!='pB10-pB10')]\n",
    "\n",
    "plt.rcParams.update({\n",
    "    \"text.usetex\": False,\n",
    "    \"font.family\": \"Helvetica\"\n",
    "})\n",
    "#Figure 1\n",
    "fig, (ax1, ax2) = plt.subplots(nrows=1, ncols=2, figsize=(7,2.7), dpi=300)\n",
    "sns.set(style='ticks')\n",
    "\n",
    "\n",
    "figure1data['plate_counts'] = [1.54e6, 9.51e5,1.52e6,1.87e6,1.24e6,0,1.73e8,1.18e7,1.37e6,1.94e5,1.44e4,0]\n",
    "#figure1data['plate_counts'] = [7.8e4, 1.17e5,1.40e5,4.67e5,3.11e5,0,1.12e7,6.83e5,5.55e4,2.66e4,1.05e3,0]\n",
    "figure1data['bacteria'] = ['Donor']*6 + ['Transconjugant']*6\n",
    "sns.barplot(data = figure1data, x='Index', y='plate_counts', \n",
    "                      hue='bacteria', palette = 'bright', ax=ax1)\n",
    "ax1.set_yscale('log')\n",
    "ax1.set_ylabel('Bacteria (CFU/g)', fontsize=7, fontweight='bold')\n",
    "ax1.set_xlabel('Soil Microcosm', fontsize=7, fontweight='bold')\n",
    "ax1.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax1.set_ylim(1, 1000000000)\n",
    "plt.minorticks_off()\n",
    "ax1.legend(loc='upper right', bbox_to_anchor=(1.58,1.23), ncol=2)\n",
    "plt.setp(ax1.get_legend().get_texts(), fontsize='7') \n",
    "plt.setp(ax1.get_legend().get_title(), fontsize='7') \n",
    "plt.setp(ax1.spines.values(), linewidth=.75)\n",
    "ax1.xaxis.set_tick_params(width=.75)\n",
    "ax1.yaxis.set_tick_params(width=.75)\n",
    "\n",
    "sns.barplot(data = figure1data, x='Index', y='count', \n",
    "            hue='Hi-C link', palette = 'bright', ax=ax2)\n",
    "ax2.set_yscale('log')\n",
    "ax2.set_ylabel('Hi-C Reads (Standardized Count)', fontsize=7, fontweight='bold')\n",
    "ax2.set_xlabel('Soil Microcosm', fontsize=7, fontweight='bold')\n",
    "ax2.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax2.set_ylim(1, 10000)\n",
    "plt.minorticks_off()\n",
    "plt.legend('', frameon=False)\n",
    "plt.setp(ax2.get_legend().get_texts(), fontsize='7') \n",
    "plt.setp(ax2.get_legend().get_title(), fontsize='7') \n",
    "plt.setp(ax2.spines.values(), linewidth=.75)\n",
    "ax2.xaxis.set_tick_params(width=.75)\n",
    "ax2.yaxis.set_tick_params(width=.75)\n",
    "\n",
    "plt.savefig('Fig2.pdf', bbox_inches='tight', dpi=300)\n",
    "plt.savefig('Fig2.png', bbox_inches='tight', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plate_table = all_data[(all_data['type']=='Hi-C+')&(all_data['Hi-C link']!='pB10-pB10')]\n",
    "plate_table['plate_counts'] = [9.51e5,1.52e6,1.87e6,1.24e6,9.27e5,0,0,1.18e7,1.37e6,1.94e5,1.44e4,1.32e3,0,0]\n",
    "plate_table['bacteria'] = ['Donor']*7 + ['Transconjugant']*7\n",
    "plate_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(7, 7), dpi=300)\n",
    "gs = fig3.add_gridspec(100, 100)\n",
    "\n",
    "rc = {'font.size': 7.0, 'axes.labelsize': 7.0, 'axes.titlesize': 7.0, 'xtick.labelsize': 7.0,\n",
    "      'ytick.labelsize': 7.0, 'legend.fontsize': 7.0, 'legend.title_fontsize': 7.0, \"font.family\": \"Helvetica\"}\n",
    "sns.set(style='ticks', rc=rc)\n",
    "\n",
    "ax1 = fig.add_subplot(gs[:30,0:30])\n",
    "ax2 = fig.add_subplot(gs[:30,35:65])\n",
    "ax3 = fig.add_subplot(gs[:30,70:])\n",
    "ax4 = fig.add_subplot(gs[40:68,0:30])\n",
    "ax5 = fig.add_subplot(gs[40:68,35:65])\n",
    "ax6 = fig.add_subplot(gs[40:68,70:])\n",
    "ax7 = fig.add_subplot(gs[72:,0:30])\n",
    "ax8 = fig.add_subplot(gs[72:,35:65])\n",
    "ax9 = fig.add_subplot(gs[72:,70:])\n",
    "plt.figtext(0.02, 0.89, 'A)', fontsize=10)\n",
    "plt.figtext(0.02, 0.57, 'B)', fontsize=10)\n",
    "all_data = all_data[all_data['Index']!='Ctrl2']\n",
    "#----------------------------------- panel A -----------------------------------------------\n",
    "sns.barplot(data = all_data[all_data['Hi-C link']=='pB10-pB10'], x='Index', y='count', \n",
    "            hue='type', palette={'Hi-C':'red', 'Hi-C+':'tan'}, ax=ax1, \n",
    "            order=['B', 'C', 'D', 'E', 'F', 'Ctrl1']).set(yscale='log')\n",
    "ax1.set_ylim(1, 100000000)\n",
    "ax1.set_ylabel('Hi-C Reads (Standardized Count)', fontsize=7, fontweight='bold')\n",
    "ax1.set(xlabel=None)\n",
    "ax1.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax1.set_title('Hi-C Read: pB10-pB10', fontsize=7, fontweight='bold')\n",
    "ax1.legend('',frameon=False)\n",
    "\n",
    "sns.barplot(data = all_data[all_data['Hi-C link']=='Ecoli-pB10'], x='Index', y='count', \n",
    "            hue='type', palette={'Hi-C':'red', 'Hi-C+':'tan'}, ax=ax2, \n",
    "            order=['B', 'C', 'D', 'E', 'F', 'Ctrl1']).set(yscale='log')\n",
    "ax2.set_xlabel('Soil Microcosm', fontsize=7, fontweight='bold')\n",
    "ax2.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax2.legend(loc='upper center', ncol=2)\n",
    "ax2.set(ylabel=None)\n",
    "ax2.set_title('Hi-C Read: Ecoli-pB10', fontsize=7, fontweight='bold')\n",
    "ax2.set_ylim(1, 100000000)\n",
    "ax2.tick_params(axis='y', label1On=False)\n",
    "ax2.yaxis.set_ticklabels([])\n",
    "\n",
    "sns.barplot(data = all_data[all_data['Hi-C link']=='Pputida-pB10'], x='Index', y='count', \n",
    "            hue='type', palette={'Hi-C':'red', 'Hi-C+':'tan'}, ax=ax3, \n",
    "            order=['B', 'C', 'D', 'E', 'F', 'Ctrl1']).set(yscale='log')\n",
    "ax3.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax3.set(ylabel=None)\n",
    "ax3.set(xlabel=None)\n",
    "ax3.set_title('Hi-C Read: Pputida-pB10', fontsize=7, fontweight='bold')\n",
    "ax3.set_ylim(1, 100000000)\n",
    "ax3.tick_params(axis='y', label1On=False)\n",
    "ax3.yaxis.set_ticklabels([])\n",
    "ax3.legend('',frameon=False)\n",
    "\n",
    "#--------------------------------------panel B------------------------------------------------\n",
    "ax4.plot(x, Bplus_all[1][1],\n",
    "        color='black',\n",
    "        linewidth=.05)\n",
    "ax4.fill_between(x, Bplus_all[1][1], 0, color='black', alpha=.5)\n",
    "ax4.set(yscale='log')\n",
    "ax4.set_ylabel('B+ Coverage', fontsize=7, fontweight='bold')\n",
    "ax4.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax4.autoscale(enable=True, axis='x', tight=True)\n",
    "ax4.margins(y=0)\n",
    "plt.setp(ax4.spines.values(), linewidth=.75)\n",
    "ax4.xaxis.set_tick_params(width=.75)\n",
    "ax4.yaxis.set_tick_params(width=.75)\n",
    "ax4.set_ylim(1, 1000000)\n",
    "ax4.xaxis.set_ticklabels([])\n",
    "ax4.minorticks_off()\n",
    "\n",
    "ax5.plot(x, Bplus_all[1][0],\n",
    "        color='black',\n",
    "        linewidth=.1)\n",
    "ax5.fill_between(x, Bplus_all[1][0], 0, color='black', alpha=.5)\n",
    "ax5.set(yscale='log')\n",
    "ax5.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax5.autoscale(enable=True, axis='x', tight=True)\n",
    "ax5.margins(y=0)\n",
    "plt.setp(ax5.spines.values(), linewidth=.75)\n",
    "ax5.xaxis.set_tick_params(width=.75)\n",
    "ax5.yaxis.set_tick_params(width=.75) \n",
    "ax5.set_ylim(1, 1000000)\n",
    "ax5.yaxis.set_ticklabels([])\n",
    "ax5.minorticks_off()\n",
    "ax5.xaxis.set_ticklabels([])\n",
    "\n",
    "\n",
    "ax6.plot(x, Bplus_all[1][2],\n",
    "        color='black',\n",
    "        linewidth=.1)\n",
    "ax6.fill_between(x, Bplus_all[1][2], 0, color='black', alpha=.5)\n",
    "ax6.set(yscale='log')\n",
    "ax6.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax6.autoscale(enable=True, axis='x', tight=True)\n",
    "ax6.margins(y=0)\n",
    "plt.setp(ax6.spines.values(), linewidth=.75)\n",
    "ax6.xaxis.set_tick_params(width=.75)\n",
    "ax6.yaxis.set_tick_params(width=.75)\n",
    "ax6.set_ylim(1, 1000000)\n",
    "ax6.yaxis.set_ticklabels([])\n",
    "ax6.minorticks_off()\n",
    "ax6.xaxis.set_ticklabels([])\n",
    "\n",
    "\n",
    "ax7.plot(x, ctrl1plus_all[1][1],\n",
    "        color='green',\n",
    "        linewidth=.1,)\n",
    "ax7.fill_between(x, ctrl1plus_all[1][1], 0, color='green', alpha=.5)\n",
    "ax7.set_ylabel('Ctrl+ Coverage', fontsize=7, fontweight='bold')\n",
    "ax7.set(yscale='log')\n",
    "ax7.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax7.autoscale(enable=True, axis='x', tight=True)\n",
    "ax7.margins(y=0)\n",
    "plt.setp(ax7.spines.values(), linewidth=.75)\n",
    "ax7.xaxis.set_tick_params(width=.75)\n",
    "ax7.yaxis.set_tick_params(width=.75) \n",
    "ax7.set_ylim(1, 1000000)\n",
    "ax7.minorticks_off()\n",
    "\n",
    "\n",
    "ax8.plot(x, ctrl1plus_all[1][0],\n",
    "        color='green',\n",
    "        linewidth=.1,)\n",
    "ax8.fill_between(x, ctrl1plus_all[1][0], 0, color='green', alpha=.5)\n",
    "ax8.set(yscale='log')\n",
    "ax8.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax8.autoscale(enable=True, axis='x', tight=True)\n",
    "ax8.margins(y=0)\n",
    "plt.setp(ax8.spines.values(), linewidth=.75)\n",
    "ax8.xaxis.set_tick_params(width=.75)\n",
    "ax8.yaxis.set_tick_params(width=.75)\n",
    "ax8.set_ylim(1, 1000000)\n",
    "ax8.set_xlabel('pB10::rfp genome coordinates', fontsize=7, fontweight='bold')\n",
    "ax8.yaxis.set_ticklabels([])\n",
    "ax8.minorticks_off()\n",
    "\n",
    "ax9.plot(x, ctrl1plus_all[1][2],\n",
    "        color='green',\n",
    "        linewidth=.1,)\n",
    "ax9.fill_between(x, ctrl1plus_all[1][2], 0, color='green', alpha=.5)\n",
    "ax9.set(yscale='log')\n",
    "ax9.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax9.autoscale(enable=True, axis='x', tight=True)\n",
    "ax9.margins(y=0)\n",
    "plt.setp(ax9.spines.values(), linewidth=.75)\n",
    "ax9.xaxis.set_tick_params(width=.75)\n",
    "ax9.yaxis.set_tick_params(width=.75)\n",
    "ax9.set_ylim(1, 1000000)\n",
    "ax9.yaxis.set_ticklabels([])\n",
    "ax9.minorticks_off()\n",
    "\n",
    "plt.savefig('Fig3.pdf', bbox_inches='tight', dpi=300)\n",
    "plt.savefig('Fig3.png', bbox_inches='tight', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/mnt/ceph/cast9836/00_projects/hic_targetcapture/06_output/figures/')\n",
    "libs = [Bplus_all, Cplus_all, Dplus_all, Eplus_all, Fplus_all, ctrl1plus_all]\n",
    "indexes = ['B', 'C', 'D', 'E', 'F', 'Ctrl1']\n",
    "\n",
    "plot_cov = pd.DataFrame()\n",
    "\n",
    "def unique_positions(ctrl, pb10):\n",
    "    unique = []\n",
    "    for i, j in zip(ctrl, pb10):\n",
    "        if i == 0 and j > 0:\n",
    "            unique.append(1)\n",
    "        else:\n",
    "            unique.append(0)\n",
    "    return(unique)\n",
    "\n",
    "for i in range(len(libs)):\n",
    "    all_ctrl1 = list(map(add,ctrl1plus_all[1][0],ctrl1plus_all[1][1]))\n",
    "    all_ctrl2 = list(map(add, all_ctrl1, ctrl1plus_all[1][2]))\n",
    "    all_ctrl3 = list(map(add, all_ctrl2, ctrl2plus_all[1][0]))\n",
    "    all_ctrl4 = list(map(add, all_ctrl3, ctrl2plus_all[1][1]))\n",
    "    all_ctrl = list(map(add, all_ctrl4, ctrl2plus_all[1][2]))\n",
    "    \n",
    "    df1 = pd.DataFrame.from_dict({'covered_bases':np.count_nonzero(libs[i][1][0]),\n",
    "                                  'unique_bases':np.count_nonzero(unique_positions(all_ctrl, libs[i][1][0])),\n",
    "                                  'percent_detected':(np.count_nonzero(libs[i][1][0])/len(libs[i][1][0])*100),\n",
    "                                  'link':['Pputida-pB10']})\n",
    "    df2 = pd.DataFrame.from_dict({'covered_bases':np.count_nonzero(libs[i][1][1]),\n",
    "                                  'unique_bases':np.count_nonzero(unique_positions(all_ctrl, libs[i][1][1])),\n",
    "                                  'percent_detected':(np.count_nonzero(libs[i][1][1])/len(libs[i][1][1])*100),\n",
    "                                  'link':['pB10-pB10']})\n",
    "    df3 = pd.DataFrame.from_dict({'covered_bases':np.count_nonzero(libs[i][1][2]), \n",
    "                                  'unique_bases':np.count_nonzero(unique_positions(all_ctrl, libs[i][1][2])),\n",
    "                                  'percent_detected':(np.count_nonzero(libs[i][1][2])/len(libs[i][1][2])*100),\n",
    "                                  'link':['Ecoli-pB10']})\n",
    "    comp_df = pd.concat([df2, df3,df1])\n",
    "    comp_df['Indexes'] = indexes[i]\n",
    "    plot_cov = pd.concat([plot_cov, comp_df])\n",
    "plot_cov['percent_unique'] = plot_cov['unique_bases']/len(Bplus_all[1][1])*100\n",
    "plot_ready = pd.DataFrame()\n",
    "plot_ready['link'] = list(plot_cov['link'])*2\n",
    "plot_ready['percent'] = list(plot_cov['percent_detected']) + list(plot_cov['percent_unique'])\n",
    "plot_ready['Indexes'] = list(plot_cov['Indexes'])*2\n",
    "plot_ready['type'] = ['Total Bases Detected']*18 + ['Unique Bases Detected']*18\n",
    "\n",
    "fig = plt.figure(figsize=(7, 2), dpi=300)\n",
    "gs = fig3.add_gridspec(100, 100)\n",
    "\n",
    "rc = {'font.size': 7.0, 'axes.labelsize': 7.0, 'axes.titlesize': 7.0, 'xtick.labelsize': 7.0,\n",
    "      'ytick.labelsize': 7.0, 'legend.fontsize': 7.0, 'legend.title_fontsize': 7.0, \"font.family\": \"Helvetica\"}\n",
    "sns.set(style='ticks', rc=rc)\n",
    "\n",
    "ax1 = fig.add_subplot(gs[:,0:30])\n",
    "ax2 = fig.add_subplot(gs[:,35:65])\n",
    "ax3 = fig.add_subplot(gs[:,70:])\n",
    "\n",
    "#left panel --------------------------------------------------------------------------\n",
    "sns.barplot(data = plot_ready[plot_ready['link']=='pB10-pB10'], x='Indexes', y='percent', \n",
    "            hue='type', ax=ax1, \n",
    "            order=['B', 'C', 'D', 'E', 'F', 'Ctrl1'])\n",
    "ax1.set_ylim(1, 100)\n",
    "ax1.set_ylabel('Percentage', fontsize=7, fontweight='bold')\n",
    "ax1.set_xlabel('Soil Microcosm', fontsize=7, fontweight='bold')\n",
    "ax1.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax1.set_title('Hi-C link: pB10-pB10', fontsize=7, fontweight='bold')\n",
    "ax1.legend('',frameon=False)\n",
    "\n",
    "\n",
    "#center panel ----------------------------------------------------------------------------\n",
    "sns.barplot(data = plot_ready[plot_ready['link']=='Ecoli-pB10'], x='Indexes', y='percent', \n",
    "            hue='type', ax=ax2, \n",
    "            order=['B', 'C', 'D', 'E', 'F', 'Ctrl1'])\n",
    "ax2.set_xlabel('Soil Microcosm', fontsize=7, fontweight='bold')\n",
    "ax2.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax2.set(ylabel=None)\n",
    "ax2.set_title('Hi-C link: Ecoli-pB10', fontsize=7, fontweight='bold')\n",
    "ax2.set_ylim(1, 100)\n",
    "ax2.tick_params(axis='y', label1On=False)\n",
    "ax2.set_yticks([])\n",
    "ax2.legend(loc='upper center', bbox_to_anchor=(.5,1.45), title='Genome Metric', ncol=2)\n",
    "\n",
    "\n",
    "#right panel --------------------------------------------------------------------------------\n",
    "sns.barplot(data = plot_ready[plot_ready['link']=='Pputida-pB10'], x='Indexes', y='percent', \n",
    "            hue='type', ax=ax3, \n",
    "            order=['B', 'C', 'D', 'E', 'F', 'Ctrl1'])\n",
    "ax3.set_xlabel('Soil Microcosm', fontsize=7, fontweight='bold')\n",
    "ax3.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax3.set(ylabel=None)\n",
    "ax3.set_title('Hi-C link: Pputida-pB10', fontsize=7, fontweight='bold')\n",
    "ax3.set_ylim(1, 100)\n",
    "ax3.tick_params(axis='y', label1On=False)\n",
    "ax3.set_yticks([])\n",
    "ax3.legend('',frameon=False)\n",
    "\n",
    "#plt.savefig('Fig3.pdf', bbox_inches='tight', dpi=300)\n",
    "#plt.savefig('Fig3.png', bbox_inches='tight', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/mnt/ceph/cast9836/00_projects/hic_targetcapture/06_output/figures/')\n",
    "\n",
    "suppfig2 = plt.figure(figsize=(7, 8), dpi=300)\n",
    "gs = suppfig2.add_gridspec(100, 100)\n",
    "\n",
    "rc = {'font.size': 7.0, 'axes.labelsize': 7.0, 'axes.titlesize': 7.0, 'xtick.labelsize': 7.0,\n",
    "      'ytick.labelsize': 7.0, 'legend.fontsize': 7.0, 'legend.title_fontsize': 7.0, \"font.family\": \"Helvetica\"}\n",
    "sns.set(style='ticks', rc=rc)\n",
    "\n",
    "ax1 = suppfig2.add_subplot(gs[:18,0:30])\n",
    "ax2 = suppfig2.add_subplot(gs[:18,35:65])\n",
    "ax3 = suppfig2.add_subplot(gs[:18,70:])\n",
    "\n",
    "ax4 = suppfig2.add_subplot(gs[21:38,0:30])\n",
    "ax5 = suppfig2.add_subplot(gs[21:38,35:65])\n",
    "ax6 = suppfig2.add_subplot(gs[21:38,70:])\n",
    "\n",
    "ax7 = suppfig2.add_subplot(gs[41:58,0:30])\n",
    "ax8 = suppfig2.add_subplot(gs[41:58,35:65])\n",
    "ax9 = suppfig2.add_subplot(gs[41:58,70:])\n",
    "\n",
    "ax10 = suppfig2.add_subplot(gs[61:78,0:30])\n",
    "ax11 = suppfig2.add_subplot(gs[61:78,35:65])\n",
    "ax12 = suppfig2.add_subplot(gs[61:78,70:])\n",
    "\n",
    "ax13 = suppfig2.add_subplot(gs[82:,0:30])\n",
    "ax14 = suppfig2.add_subplot(gs[82:,35:65])\n",
    "ax15 = suppfig2.add_subplot(gs[82:,70:])\n",
    "\n",
    "def plot_coverage(library, ax1, ax2, ax3, xlabels='off'):\n",
    "    x = list(range(len(library[1][0])))\n",
    "    ax1.plot(x, library[1][1],\n",
    "        color='black',\n",
    "        linewidth=.05)\n",
    "    ax1.fill_between(x, library[1][1], 0, color='black', alpha=.5)\n",
    "    \n",
    "    ax2.plot(x, library[1][2],\n",
    "            color='black',\n",
    "            linewidth=.1)    \n",
    "    ax2.fill_between(x, library[1][2], 0, color='black', alpha=.5)\n",
    "    \n",
    "    ax3.plot(x, library[1][0],\n",
    "            color='black',\n",
    "            linewidth=.1)    \n",
    "    ax3.fill_between(x, library[1][0], 0, color='black', alpha=.5)\n",
    "    \n",
    "    for ax in [ax1, ax2, ax3]:\n",
    "        ax.set(yscale='log')\n",
    "        ax.tick_params(axis='both', which='major', labelsize=7)\n",
    "        ax.autoscale(enable=True, axis='x', tight=True)\n",
    "        ax.margins(y=0)\n",
    "        plt.setp(ax1.spines.values(), linewidth=.75)\n",
    "        ax.xaxis.set_tick_params(width=.75)\n",
    "        ax.yaxis.set_tick_params(width=.75)\n",
    "        ax.set_ylim(1, 1000000)\n",
    "        ax.minorticks_off()\n",
    "        if xlabels == 'off':\n",
    "            ax.xaxis.set_ticklabels([])\n",
    "\n",
    "    ax2.yaxis.set_ticklabels([])\n",
    "    ax3.yaxis.set_ticklabels([])\n",
    "\n",
    "#make plots\n",
    "plot_coverage(Bplus_all, ax1, ax2, ax3)\n",
    "plot_coverage(Cplus_all, ax4, ax5, ax6)\n",
    "plot_coverage(Dplus_all, ax7, ax8, ax9)\n",
    "plot_coverage(Eplus_all, ax10, ax11, ax12)\n",
    "plot_coverage(Fplus_all, ax13, ax14, ax15, 'On')\n",
    "\n",
    "#add labels\n",
    "ax1.set_ylabel('B+ Coverage', fontsize=7, fontweight='bold')\n",
    "ax4.set_ylabel('C+ Coverage', fontsize=7, fontweight='bold')\n",
    "ax7.set_ylabel('D+ Coverage', fontsize=7, fontweight='bold')\n",
    "ax10.set_ylabel('E+ Coverage', fontsize=7, fontweight='bold')\n",
    "ax13.set_ylabel('F+ Coverage', fontsize=7, fontweight='bold')\n",
    "ax14.set_xlabel('pB10::rfp genome coordinates', fontsize=7, fontweight='bold')\n",
    "\n",
    "ax1.set_title('Hi-C Read: pB10-pB10', fontsize=7, fontweight='bold')\n",
    "ax2.set_title('Hi-C Read: Ecoli-pB10', fontsize=7, fontweight='bold')\n",
    "ax3.set_title('Hi-C Read: Pputida-pB10', fontsize=7, fontweight='bold')\n",
    "\n",
    "plt.savefig('supp2.pdf', bbox_inches='tight', dpi=300)\n",
    "plt.savefig('supp2.png', bbox_inches='tight', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(range(len(Bplus_all[1][0])))[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir('/mnt/ceph/cast9836/00_projects/hic_targetcapture/06_output/figures/')\n",
    "\n",
    "def extract_gene(loc_list):\n",
    "    plasmid = SeqIO.read('../../04_ref_genomes/pB10__rfp putative.gb', 'genbank')\n",
    "    location = []\n",
    "    gene = []\n",
    "    for loc in loc_list:\n",
    "        temp = []\n",
    "        temp_l = []\n",
    "        for feature in plasmid.features:\n",
    "            if feature.type == 'gene' and loc in feature:\n",
    "                if 'gene' in feature.qualifiers.keys():\n",
    "                    t = feature.qualifiers['gene']\n",
    "                    temp.append(str(t)[2:-2])\n",
    "                    temp_l.append(loc)\n",
    "                else:\n",
    "                    t = feature.qualifiers['standard_name']\n",
    "                    temp.append(str(t)[2:-2])\n",
    "                    temp_l.append(loc)\n",
    "        if len(temp_l) == 1:\n",
    "            location.append(temp_l[0])\n",
    "            gene.append(str(temp[0]))\n",
    "        elif len(temp_l) == 2:\n",
    "            location.append(temp_l[0])\n",
    "            gene.append(str(temp[0])+' and '+str(temp[1]))\n",
    "    return([location, gene])\n",
    "\n",
    "\"\"\"def conv_to_counts(linked_reads, loc_and_gene):\n",
    "    no_gene = len(linked_reads) - (len(loc_and_gene[0]))\n",
    "    all_genes = loc_and_gene[1]\n",
    "    unique_genes = list(set(loc_and_gene[1]))\n",
    "    apps = []\n",
    "    for i in unique_genes:\n",
    "        apps.append(all_genes.count(i))\n",
    "    unique_genes.append('No gene')\n",
    "    apps.append(no_gene)\n",
    "    return(unique_genes, apps)\"\"\"\n",
    "\n",
    "def conv_to_counts(loc_and_gene):\n",
    "    all_genes = loc_and_gene[1]\n",
    "    unique_genes = list(set(loc_and_gene[1]))\n",
    "    apps = []\n",
    "    for i in unique_genes:\n",
    "        apps.append(all_genes.count(i))\n",
    "    return(unique_genes, apps)\n",
    "\n",
    "ctrl1_pb_pb_genes = extract_gene(ctrl1plus_all[2][1])\n",
    "ctrl1_pb_ecoli_genes = extract_gene(ctrl1plus_all[2][2])\n",
    "ctrl1_pb_pputida_genes = extract_gene(ctrl1plus_all[2][0])\n",
    "\n",
    "ctrl1_pb_pb_counts = conv_to_counts(ctrl1_pb_pb_genes)\n",
    "ctrl1_pb_ecoli_counts = conv_to_counts(ctrl1_pb_ecoli_genes)\n",
    "ctrl1_pb_pputida_counts = conv_to_counts(ctrl1_pb_pputida_genes)\n",
    "\n",
    "ctrl1_pb_pb_df = pd.DataFrame(list(zip(ctrl1_pb_pb_counts[0],ctrl1_pb_pb_counts[1])), columns=[1,2]).explode(2).reset_index(drop=True).add_prefix('column')\n",
    "ctrl1_pb_ecoli_df = pd.DataFrame(list(zip(ctrl1_pb_ecoli_counts[0],ctrl1_pb_ecoli_counts[1])), columns=[1,2]).explode(2).reset_index(drop=True).add_prefix('column')\n",
    "ctrl1_pb_pputida_df = pd.DataFrame(list(zip(ctrl1_pb_pputida_counts[0],ctrl1_pb_pputida_counts[1])), columns=[1,2]).explode(2).reset_index(drop=True).add_prefix('column')\n",
    "\n",
    "ctrl1_pb_pb_df.loc[54] = ['tnpA', 23822+18468]\n",
    "ctrl1_pb_pb_df.drop([2,3], inplace=True)\n",
    "ctrl1_pb_pb_df.rename(columns={'column1':'Gene', 'column2':'pB10-pB10'}, inplace=True)\n",
    "ctrl1_pb_pputida_df.loc[0] = ['tnpA', 88]\n",
    "ctrl1_pb_pputida_df.rename(columns={'column1':'Gene', 'column2':'Pputida-pB10'}, inplace=True)\n",
    "ctrl1_pb_ecoli_df.rename(columns={'column1':'Gene', 'column2':'Ecoli-pB10'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df = ctrl1_pb_ecoli_df.merge(ctrl1_pb_pputida_df, on='Gene', how='outer').merge(ctrl1_pb_pb_df, on='Gene', how='outer')\n",
    "combined_df.fillna(0, inplace=True)\n",
    "combined_df = combined_df.sort_values(by=['Gene']).set_index('Gene')\n",
    "\n",
    "plt.rcParams.update({\n",
    "    \"text.usetex\": False,\n",
    "    \"font.family\": \"Helvetica\"\n",
    "})\n",
    "\n",
    "#Supplementary figure\n",
    "fig, ax1 = plt.subplots(figsize=(7,2.7), dpi=300)\n",
    "combined_df.plot.bar(stacked=True, ax=ax1)\n",
    "ax1.set(yscale='log')\n",
    "ax1.set_ylabel('Count', fontsize=7, fontweight='bold')\n",
    "ax1.set_xlabel('Gene', fontsize=7, fontweight='bold')\n",
    "ax1.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax1.legend(loc='upper center', ncol=3)\n",
    "plt.setp(ax1.get_legend().get_texts(), fontsize='7')\n",
    "ax1.set_ylim(1, 500000)\n",
    "ax1.minorticks_off()\n",
    "\n",
    "plt.savefig('supp3.pdf', bbox_inches='tight', dpi=300)\n",
    "plt.savefig('supp3.png', bbox_inches='tight', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctrl1_pb_pputida_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctrl1_pb_pputida_df.loc[0] = ['tnpA', 88]\n",
    "ctrl1_pb_pputida_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_gene = {735595: 'ybfO', 'blast search detected sequence in salmonells sp hnk130, shigella flexneri, and escherichia genus'\n",
    "              739397: 'ybgA, DUF1722 domain-containing protein YbgA, NP_415235.1, same with this one. protein found in shigella and escherichia'}\n",
    "\n",
    "pputida_genes = {5024355: 'PP_RS22990, PLP-dependent aminotransferase family protein, WP_029886623.1' 'no sim on plasmids', \n",
    "                 1818628: 'PP_RS08360, peptidoglycan DD-metalloendopeptidase family protein, WP_049588177.1' 'Pseudomonas sp. TMW 2.1634 plasmid pL21634-2, complete sequence', \n",
    "                 5689795: 'pilH, twitching motility response regulator PilH, WP_003249327.1', 'no sim', \n",
    "                 426631: 'PP_RS01860, TonB-dependent siderophore receptor, WP_010951702.1' 'no sim', \n",
    "                 595884: 'PP_RS02685, ABC transporter permease, WP_010951795.1', \n",
    "                 3405935+3405944: 'PP_RS15610, methylated-DNA--[protein]-cysteine S-methyltransferase, WP_010953916.1', \n",
    "                 6148755, 'PP_RS28095, heavy metal translocating P-type ATPase' \n",
    "                 4054047, 'PP_RS18565, NAD(P)/FAD-dependent oxidoreductase, WP_010954426.1'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctrl1plus_all[1][1]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  },
  "vscode": {
   "interpreter": {
    "hash": "1ba4d73b7b8f239b5baf417115c673058c1ed585beac6d3c39d5a9a12ecb10fd"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
